content/blog/2019-10-30-loop-reduction/index.md

```diff
@@ -0,0 +1,375 @@
+
+In implementation, for each block, we first copy old block content and then check if the next block is a loop header. If so, we create an empty block. 
+
+```python
+for edge in loops:#we use back edge as key to denote loop
+    if b_names[i+1] in edge[1]: #edge[1] is the pre-header block name
+        name = fresh('b', new_blocks) # generate block name that is never used before
+        new_blocks[name] = []
+        pre_header = {x:name for x in loops[edge]}
+        break
+```
+
+<img src="move_LI.png" alt="reaching definition lecture" style="zoom:60%;" />
+
+### Move Loop Invariant to Pre-Headers
+
+Not all pre-headers are allowed to be moved to the pre-headers. If the destination of an LI is `d`, it needs to satisfy the following condition:
+
+1. There is only one definition of  `d` in the loop
+2. `d`dominates all its uses, or equivalently, `d` is not live-out of it's pre-header.
+3. `d`s block dominates all loop exits where $d$ is live-out
+
+To learn the first condition, we need to know all definitions inside the loop and check if  `d` is unique in the list `defs`
+
+```python
+defs = [ins.get('dest') for b in loops[back_edge] for ins in blocks[b] if ins.get('dest')]
+defs.count(instr['dest']) ==  # if true, first check passed
+```
+
+For the second condition, we can check the predecessor block of pre-header we just added, by simply read the index of pre-header and subtract 1. Then check if `d` is live-out of block.
+
+```python
+ind = b_names.index(pre_header[b_name]) - 1 #b_name is the name of block where d is LI.
+instr['dest'] not in live_var[1][b_names[ind]] #if true, second check passed
+```
+
+For the third condition, we need information of exits of blocks. The exits are blocks that have successors not in the loop. 
+
+```python
+exits = {}
+for k in loops:
+    exits[k] = []
+    for l in loops[k]: 
+        if any( s not in loops[k] for s in succ[l]):
+            exits[k].append(l)
+```
+
+After that, we just need to find all exit blocks where `d` is live-out and check if `d`'s' block dominates all loop exits.
+
+```python
+edest = [e for e in exits[back_edge] if instr['dest'] in live_var[1][e]]
+all([b_name in dom[e] for e in edest]) # if true, third check passed.
+```
+
+### Block to Dictionaries
+
+` json.load(sys.stdin)['functions']` gives us dictionaries and for each dictionary we obtain list of instructions when the key is `instrs`. Then we change this list of instructions into a directory of blocks. We would like to reverse this process to regenerate list of instructions with modified blocks. However, original block does not have so many labels introduced when generating block names and creating pre-headers. Luckily, the modified blocks are still ordered dictionary and we can omit blocks we introduced. Therefore in this step, we only create `label` instruction when the original labels in blocks. The rest is just copy every instruction other than labels to the new list of instructions.
```

Comment:
For example, main function does not have a label, but to form a block, we generate a `b0` for its label. Similarly, when generating a pre-header, we might have inserted blocked labeled with `b1` .  Luckily, the modified blocks are still in an ordered dictionary. Blocks with no label at first is safe to be unlabled and for `pre-header` blocks, we assume they have only one in-edge and one out-edge.
So it is safe here to just delete `label` that are not originally there.

Reply:
Got it. It might be wort explaining this fact in your post.

For what it's worth, the reason to create pre-headers in the first place is typically that loop headers can have multiple predecessors. So pre-headers generally do too. The idea is to make it so the loop _header_ has only a single predecessor, which is a place you can put pre-loop code. I think you've just assumed that away, but I wanted to point out that this is nonstandard. If you require loop headers to have only a single predecessor, then there is no real reason to create pre-headers—you can just put the code right in that predecessor.

Reply:
Yes, I've put the explanation in the updated post. Thanks for pointing out that the previous explantion is hard to follow.
We decided to create pre-header blocks because we work on LICM and LR separately, and inserting instructions to an empty block might result in less error when we try to merging things together. And yes, the pre-header is not compulsory rather than just a design choice. I will put this line in the post either.